### Biblioteki i przygotowanie danych

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import arff

In [2]:
titanic_arff = arff.load(open("Zbiór danych Titanic.arff", 'r'))
print(titanic_arff.keys())

attribiutes = titanic_arff["attributes"]
data = titanic_arff["data"]

df = pd.DataFrame(data, columns=[x[0] for x in attribiutes])

df.head(5)

dict_keys(['description', 'relation', 'attributes', 'data'])


pclass survived                                             name     sex  \
0     1.0        1                    Allen, Miss. Elisabeth Walton  female   
1     1.0        1                   Allison, Master. Hudson Trevor    male   
2     1.0        0                     Allison, Miss. Helen Loraine  female   
3     1.0        0             Allison, Mr. Hudson Joshua Creighton    male   
4     1.0        0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked  boat   body  \
0  29.0000    0.0    0.0   24160  211.3375       B5        S     2    NaN   
1   0.9167    1.0    2.0  113781  151.5500  C22 C26        S    11    NaN   
2   2.0000    1.0    2.0  113781  151.5500  C22 C26        S  None    NaN   
3  30.0000    1.0    2.0  113781  151.5500  C22 C26        S  None  135.0   
4  25.0000    1.0    2.0  113781  151.5500  C22 C26        S  None    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

### 1. Liczebność poszczególnych etykiet dla danych zmiennych jakościowych

In [3]:
print('Liczba etykiet zmiennej pclass: {}'.format(len(df.pclass.unique())))
print('Liczba etykiet zmiennej survived: {}'.format(len(df.survived.unique())))
print('Liczba etykiet zmiennej name: {}'.format(len(df.name.unique())))
print('Liczba etykiet zmiennej sex: {}'.format(len(df.sex.unique())))
print('Liczba etykiet zmiennej ticket: {}'.format(len(df.ticket.unique())))
print('Liczba etykiet zmiennej cabin: {}'.format(len(df.cabin.unique())))
print('Liczba etykiet zmiennej embarked: {}'.format(len(df.embarked.unique())))
print('Liczba etykiet zmiennej boat: {}'.format(len(df.boat.unique())))
print('Liczba etykiet zmiennej home.dest: {}'.format(len(df["home.dest"].unique())))

Liczba etykiet zmiennej pclass: 3
Liczba etykiet zmiennej survived: 2
Liczba etykiet zmiennej name: 1307
Liczba etykiet zmiennej sex: 2
Liczba etykiet zmiennej ticket: 929
Liczba etykiet zmiennej cabin: 187
Liczba etykiet zmiennej embarked: 4
Liczba etykiet zmiennej boat: 28
Liczba etykiet zmiennej home.dest: 370


### 2. Liczba wszystkich pasażerów.

In [4]:
print('Liczba wszystkich pasażerów: {}'.format(len(df)))

Liczba wszystkich pasażerów: 1309


### 3. Komentarz do wyników otrzymanych w punkcie 1 i 2.

Na podstawie liczby unikalnych etykiet można podzielić zmienne jakościowe ze względu na kardynalność - moc zbioru.

##### Zmienne o **małej kardynalności**:
- pclass: 3 klasy
- survived: 2 wartości (przeżyl/ nie przeżył)
- sex: 2 wartości plci
- embarked: 4 porty
- boat: 28 łodzi

##### Zmienne o **dużej kardynalności**:
- name: 1307 unikalnych imion i nazwisk (niemal wszyscy mają inne)
- ticket: 929 różnych numerów biletów
- cabin: 187 różnych oznaczeń kabin
- home.dest: 370 miejsc docelowych


Liczba wszystkich pasażerów: 1309. A widzimy, że kolumna name ma 1307 unikalnych etykiet, co oznacza, że niemal każda osoba w zbiorze jest unikalna.


==WNIOSKI==
- Zmienne o małej kardynalności są prostsze do zakodowania i analizy.
- Zmienne o dużej liczbie etykiet mogą być trudniejsze do przetworzenia lub modelowania – szczególnie jeśli są jakościowe. W analizie warto rozważyć ich redukcję lub transformację, aby ułatwić dalsze kroki.


### 4. Ile unikalnych etykiet ma zmienna mówiąca o kabinie danego pasażera?

In [5]:
unique_cabins = df['cabin'].unique()
print('Liczba unikalnych kabin:', len(unique_cabins))
print('Wartości (postać NumPy array):', np.array(unique_cabins))

Liczba unikalnych kabin: 187
Wartości (postać NumPy array): ['B5' 'C22 C26' 'E12' 'D7' 'A36' 'C101' None 'C62 C64' 'B35' 'A23'
 'B58 B60' 'D15' 'C6' 'D35' 'C148' 'C97' 'B49' 'C99' 'C52' 'T' 'A31' 'C7'
 'C103' 'D22' 'E33' 'A21' 'B10' 'B4' 'E40' 'B38' 'E24' 'B51 B53 B55'
 'B96 B98' 'C46' 'E31' 'E8' 'B61' 'B77' 'A9' 'C89' 'A14' 'E58' 'E49' 'E52'
 'E45' 'B22' 'B26' 'C85' 'E17' 'B71' 'B20' 'A34' 'C86' 'A16' 'A20' 'A18'
 'C54' 'C45' 'D20' 'A29' 'C95' 'E25' 'C111' 'C23 C25 C27' 'E36' 'D34'
 'D40' 'B39' 'B41' 'B102' 'C123' 'E63' 'C130' 'B86' 'C92' 'A5' 'C51' 'B42'
 'C91' 'C125' 'D10 D12' 'B82 B84' 'E50' 'D33' 'C83' 'B94' 'D49' 'D45'
 'B69' 'B11' 'E46' 'C39' 'B18' 'D11' 'C93' 'B28' 'C49' 'B52 B54 B56' 'E60'
 'C132' 'B37' 'D21' 'D19' 'C124' 'D17' 'B101' 'D28' 'D6' 'D9' 'B80' 'C106'
 'B79' 'C47' 'D30' 'C90' 'E38' 'C78' 'C30' 'C118' 'D36' 'D48' 'D47' 'C105'
 'B36' 'B30' 'D43' 'B24' 'C2' 'C65' 'B73' 'C104' 'C110' 'C50' 'B3' 'A24'
 'A32' 'A11' 'A10' 'B57 B59 B63 B66' 'C28' 'E44' 'A26' 'A6' 'A7' 'C31

Jak widzimy występuje etykieta "None" - to znaczy, że niektóre osoby nie miały przypisanej do siebie żadnej kabiny albo informacja o tym została zgubiona. Dlatego też w kolejnym zadaniu będzie pojawiać się literka N, jako pierwsza litera - sugerująca brak przypisanej kabiny

### 5. Zastąpienie obecnych etykiet w formacie LL11 do etykiet zawierających tylko pierwszą literę.

In [6]:
df['CabinReduced'] = df['cabin'].astype(str).str[0]
df[['cabin', 'CabinReduced']].head(20)

cabin CabinReduced
0        B5            B
1   C22 C26            C
2   C22 C26            C
3   C22 C26            C
4   C22 C26            C
5       E12            E
6        D7            D
7       A36            A
8      C101            C
9      None            N
10  C62 C64            C
11  C62 C64            C
12      B35            B
13     None            N
14      A23            A
15     None            N
16  B58 B60            B
17  B58 B60            B
18      D15            D
19       C6            C

### 6. Liczba etykiet dla zmiennych z pkt 5.

In [7]:
original_cardinality = len(df['cabin'].unique())
reduced_cardinality = len(df['CabinReduced'].unique())

print('Liczba etykiet przed redukcją:', original_cardinality)
print('Liczba etykiet po redukcji:', reduced_cardinality)

reduction_percent = 100 * (original_cardinality - reduced_cardinality) / original_cardinality
print('Procent redukcji kardynalności: {:.2f}%'.format(reduction_percent))


Liczba etykiet przed redukcją: 187
Liczba etykiet po redukcji: 9
Procent redukcji kardynalności: 95.19%


### 7. Dlaczego dokonuję redukcji akurat tej zmiennej? Jak to wpływa na przyszłe analizy, czy powoduje jakieś negatywne skutki?

Zmienna cabin zawiera bardzo wiele unikalnych wartości (np. D15, C62, A36.), co utrudniałoby ich kodowanie oraz zwiększałoby wymiarowość danych.

Redukując zmienną cabin do pierwszej litery (np. C101 → C), uzyskujemy mniej kategorii, które często odpowiadają zapewne **pokładom statku**. Dzięki temu:
- zmniejszamy kardynalność,
- zachowujemy istotną informację o położeniu pasażera,
- ułatwiamy analizę i modelowanie

W takiej sytuacji tracimy precyzję (różnicę między np. C101 a C64), ale w zamian otrzymujemy bardziej stabilną zmienną do analizy. W większości przypadków korzyści przeważają nad negatywami.

In [8]:
display(df.head(20))

pclass survived                                               name  \
0      1.0        1                      Allen, Miss. Elisabeth Walton   
1      1.0        1                     Allison, Master. Hudson Trevor   
2      1.0        0                       Allison, Miss. Helen Loraine   
3      1.0        0               Allison, Mr. Hudson Joshua Creighton   
4      1.0        0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5      1.0        1                                Anderson, Mr. Harry   
6      1.0        1                  Andrews, Miss. Kornelia Theodosia   
7      1.0        0                             Andrews, Mr. Thomas Jr   
8      1.0        1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9      1.0        0                            Artagaveytia, Mr. Ramon   
10     1.0        0                             Astor, Col. John Jacob   
11     1.0        1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12     1.0        1                      Aubart, Mme. Leontine Pauline   
13     1.0        1                       Barber, Miss. Ellen 'Nellie'   
14     1.0        1               Barkworth, Mr. Algernon Henry Wilson   
15     1.0        0                                Baumann, Mr. John D   
16     1.0        0                           Baxter, Mr. Quigg Edmond   
17     1.0        1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18     1.0        1                              Bazzani, Miss. Albina   
19     1.0        0                               Beattie, Mr. Thomson   

       sex      age  sibsp  parch    ticket      fare    cabin embarked  boat  \
0   female  29.0000    0.0    0.0     24160  211.3375       B5        S     2   
1     male   0.9167    1.0    2.0    113781  151.5500  C22 C26        S    11   
2   female   2.0000    1.0    2.0    113781  151.5500  C22 C26        S  None   
3     male  30.0000    1.0    2.0    113781  151.5500  C22 C26        S  None   
4   female  25.0000    1.0    2.0    113781  151.5500  C22 C26        S  None   
5     male  48.0000    0.0    0.0     19952   26.5500      E12        S     3   
6   female  63.0000    1.0    0.0     13502   77.9583       D7        S    10   
7     male  39.0000    0.0    0.0    112050    0.0000      A36        S  None   
8   female  53.0000    2.0    0.0     11769   51.4792     C101        S     D   
9     male  71.0000    0.0    0.0  PC 17609   49.5042     None        C  None   
10    male  47.0000    1.0    0.0  PC 17757  227.5250  C62 C64        C  None   
11  female  18.0000    1.0    0.0  PC 17757  227.5250  C62 C64        C     4   
12  female  24.0000    0.0    0.0  PC 17477   69.3000      B35        C     9   
13  female  26.0000    0.0    0.0     19877   78.8500     None        S     6   
14    male  80.0000    0.0    0.0     27042   30.0000      A23        S     B   
15    male      NaN    0.0    0.0  PC 17318   25.9250     None        S  None   
16    male  24.0000    0.0    1.0  PC 17558  247.5208  B58 B60        C  None   
17  female  50.0000    0.0    1.0  PC 17558  247.5208  B58 B60        C     6   
18  female  32.0000    0.0    0.0     11813   76.2917      D15        C     8   
19    male  36.0000    0.0    0.0     13050   75.2417       C6        C     A   

     body                        home.dest CabinReduced  
0     NaN                     St Louis, MO            B  
1     NaN  Montreal, PQ / Chesterville, ON            C  
2     NaN  Montreal, PQ / Chesterville, ON            C  
3   135.0  Montreal, PQ / Chesterville, ON            C  
4     NaN  Montreal, PQ / Chesterville, ON            C  
5     NaN                     New York, NY            E  
6     NaN                       Hudson, NY            D  
7     NaN                      Belfast, NI            A  
8     NaN              Bayside, Queens, NY            C  
9    22.0              Montevideo, Uruguay            N  
10  124.0                     New York, NY            C  
11    NaN                     New York, NY         

In [9]:
display(df.head(2))

pclass survived                            name     sex      age  sibsp  \
0     1.0        1   Allen, Miss. Elisabeth Walton  female  29.0000    0.0   
1     1.0        1  Allison, Master. Hudson Trevor    male   0.9167    1.0   

   parch  ticket      fare    cabin embarked boat  body  \
0    0.0   24160  211.3375       B5        S    2   NaN   
1    2.0  113781  151.5500  C22 C26        S   11   NaN   

                         home.dest CabinReduced  
0                     St Louis, MO            B  
1  Montreal, PQ / Chesterville, ON            C

In [10]:
df.to_csv('Zbior_danych_Titanic_reduced.csv', index=False)